File: frontend/lib/models/message.dart
import 'user.dart';

enum MessageType { text, gift, system }

extension MessageTypeX on MessageType {
  static MessageType fromApi(dynamic raw) {
    final value = raw?.toString().toUpperCase();
    switch (value) {
      case 'GIFT':
        return MessageType.gift;
      case 'SYSTEM':
        return MessageType.system;
      case 'TEXT':
      default:
        return MessageType.text;
    }
  }
}

class Message {
  final int id;
  final int matchId;
  final User sender;
  final User receiver;
  final String content;
  final bool isRead;
  final DateTime? readAt;
  final DateTime sentAt;
  final MessageType messageType;
  final int? referenceId;

  Message({
    required this.id,
    required this.matchId,
    required this.sender,
    required this.receiver,
    required this.content,
    this.isRead = false,
    this.readAt,
    required this.sentAt,
    this.messageType = MessageType.text,
    this.referenceId,
  });

  DateTime get timestamp => sentAt;

  factory Message.fromJson(Map<String, dynamic> json) {
    final rawMatch = json['match'];
    final matchId = rawMatch is Map ? rawMatch['id'] : json['matchId'];
    final parsedMatchId = matchId is int ? matchId : int.tryParse(matchId?.toString() ?? '') ?? 0;

    return Message(
      id: json['id'],
      matchId: parsedMatchId,
      sender: User.fromJson(Map<String, dynamic>.from(json['sender'] as Map)),
      receiver: User.fromJson(Map<String, dynamic>.from(json['receiver'] as Map)),
      content: json['content'] ?? '',
      isRead: json['isRead'] ?? false,
      readAt: json['readAt'] != null ? DateTime.parse(json['readAt']) : null,
      sentAt: DateTime.parse(json['sentAt']),
      messageType: MessageTypeX.fromApi(json['messageType']),
      referenceId: json['referenceId'] is int
          ? json['referenceId']
          : int.tryParse(json['referenceId']?.toString() ?? ''),
    );
  }
}

File: backend/src/main/java/com/trish/model/Message.java
package com.trish.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "messages")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Message {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "match_id", nullable = false)
    private Match match;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sender_id", nullable = false)
    private User sender;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "receiver_id", nullable = false)
    private User receiver;

    @Column(nullable = false, length = 2000)
    private String content;

    @Column(nullable = false)
    private Boolean isRead = false;

    @Column
    private LocalDateTime readAt;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime sentAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MessageType messageType = MessageType.TEXT;

    @Column
    private Long referenceId; // For GIFT transactions or other related entities

    public enum MessageType {
        TEXT, GIFT, SYSTEM
    }
}
